File: sdk/python/scripts/verify_contract.py
"""Check the Python SDK's wire assumptions against the server contract.

The server owns the API shape, while this client reaches into a deliberately
small part of it by hand. This check keeps those assumptions close to the
client so a breaking server change fails before it reaches an SDK user.
"""

from __future__ import annotations

import json
from pathlib import Path
import sys
from typing import Any, Dict, List


SDK = "python"
_CONTRACT_DIR = Path(__file__).resolve().parents[2] / "contract"
_CONTRACT_PATH = _CONTRACT_DIR / "contract.json"
_MANIFEST_PATH = _CONTRACT_DIR / "manifests" / (SDK + ".json")


def _read_json(path: Path) -> Dict[str, Any]:
    with path.open(encoding="utf-8") as handle:
        value = json.load(handle)
    if not isinstance(value, dict):
        raise ValueError("the top-level JSON value is not an object")
    return value


def _format_problem(scenario: str, detail: str) -> str:
    return "  %s\n      %s" % (scenario, detail)


def _check(contract: Dict[str, Any], manifest: Dict[str, Any]) -> List[str]:
    problems: List[str] = []

    def fail(scenario: str, detail: str) -> None:
        problems.append(_format_problem(scenario, detail))

    operations = contract.get("operations", {})
    for operation in manifest.get("operations", []):
        if operation not in operations:
            fail(
                "operation %s" % operation,
                "the API no longer serves it. Update the client, or drop it from the manifest.",
            )

    schemas = contract.get("schemas", {})
    for name, declared in manifest.get("schemas", {}).items():
        if name not in schemas:
            fail("schema %s" % name, "the API no longer defines it.")
            continue

        schema = schemas[name]
        properties = schema.get("properties", {})

        def check(field: str, expectation: str) -> None:
            if field not in properties:
                fail(
                    "%s.%s" % (name, field),
                    "the API no longer has this property. It has: %s"
                    % ", ".join(sorted(properties)),
                )
                return

            prop = properties[field]
            if expectation == "required" and prop.get("required") is not True:
                fail(
                    "%s.%s" % (name, field),
                    "this client reads it as always present, but the API no longer requires it.",
                )
            if expectation == "optional" and prop.get("required") is True:
                fail(
                    "%s.%s" % (name, field),
                    "this client omits it, but the API now requires it.",
                )

        for field in declared.get("required", []):
            check(field, "required")
        for field in declared.get("optional", []):
            check(field, "optional")
        for field in declared.get("fields", []):
            check(field, "present")

    for path, declared in manifest.get("enums", {}).items():
        name, field = path.split(".", 1)
        schema = schemas.get(name)
        properties = schema.get("properties", {}) if schema else {}
        if field not in properties:
            fail("enum %s" % path, "the API no longer has this property.")
            continue

        accepted = properties[field].get("enum")
        if not isinstance(accepted, list):
            fail(
                "enum %s" % path,
                "the API no longer constrains this property to an enum.",
            )
            continue

        values = declared if isinstance(declared, list) else declared.get("values", [])
        missing = [value for value in values if value not in accepted]
        if missing:
            fail(
                "enum %s" % path,
                "this client handles %s, which the API no longer accepts. It now accepts: %s"
                % (
                    ", ".join(str(value) for value in missing),
                    ", ".join(str(value) for value in accepted),
                ),
            )
        if not isinstance(declared, list) and declared.get("exhaustive") is True:
            extra = [value for value in accepted if value not in values]
            if extra:
                fail(
                    "enum %s" % path,
                    "this client claims to handle every value but the API added: %s"
                    % ", ".join(str(value) for value in extra),
                )

    return problems


def check_contract() -> List[str]:
    """Return every mismatch between this SDK's manifest and the contract."""

    return _check(_read_json(_CONTRACT_PATH), _read_json(_MANIFEST_PATH))


def main() -> int:
    try:
        contract = _read_json(_CONTRACT_PATH)
    except (OSError, UnicodeError, json.JSONDecodeError, ValueError) as error:
        print(
            "error: cannot read the wire contract (%s): %s"
            % (_CONTRACT_PATH, error),
            file=sys.stderr,
        )
        return 1
    try:
        manifest = _read_json(_MANIFEST_PATH)
    except (OSError, UnicodeError, json.JSONDecodeError, ValueError) as error:
        print(
            "error: cannot read the SDK manifest (%s): %s"
            % (_MANIFEST_PATH, error),
            file=sys.stderr,
        )
        return 1

    problems = _check(contract, manifest)
    if problems:
        print(
            "SDK contract check FAILED for %s (%s problems)\n"
            % (SDK, len(problems)),
            file=sys.stderr,
        )
        for problem in problems:
            print(problem + "\n", file=sys.stderr)
        print(
            "The server's wire contract moved. Update sdk/python to match, then\n"
            "adjust sdk/contract/manifests/python.json. See sdk/contract/README.md.",
            file=sys.stderr,
        )
        return 1

    counts = "%s operations, %s schemas, %s enums" % (
        len(manifest.get("operations", [])),
        len(manifest.get("schemas", {})),
        len(manifest.get("enums", {})),
    )
    print("SDK contract check ok for %s (%s)" % (SDK, counts))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: sdk/typescript/scripts/verify-contract.mjs
#!/usr/bin/env node
/**
 * Check this SDK's declared wire dependencies against the server's contract.
 *
 * `sdk/contract/contract.json` is the projection of the server's OpenAPI
 * document; `sdk/contract/manifests/typescript.json` is what this client says
 * it depends on. Everything named there must still exist, still be required if
 * this client relies on it being present, still be optional if this client
 * relies on being able to omit it, and still accept every enum value this
 * client sends or switches on.
 *
 * The generated types in `src/generated/openapi.ts` cover the whole document
 * and `npm run generate` proves they are current. This is the other half: the
 * hand-written layer above them — `run`, the resources, the turn follower —
 * reaches for wire fields by name, and no type check catches a field that
 * moved out from under a template string.
 *
 * The same checks run for Python, Swift and Elixir from their own manifests.
 * Keep the four implementations in step; `sdk/contract/README.md` is the spec.
 */
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const SDK = "typescript";
const here = dirname(fileURLToPath(import.meta.url));
const contractDir = join(here, "..", "..", "contract");

const read = (path, what) => {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    console.error(`error: cannot read ${what} (${path}): ${error.message}`);
    process.exit(1);
  }
};

const contract = read(join(contractDir, "contract.json"), "the wire contract");
const manifest = read(join(contractDir, "manifests", `${SDK}.json`), "the SDK manifest");

const problems = [];
const fail = (scenario, detail) => problems.push({ scenario, detail });

// ── operations ───────────────────────────────────────────────────────────────

for (const operation of manifest.operations ?? []) {
  if (!(operation in contract.operations)) {
    fail(
      `operation ${operation}`,
      "the API no longer serves it. Update the client, or drop it from the manifest.",
    );
  }
}

// ── schemas and their fields ─────────────────────────────────────────────────

for (const [name, declared] of Object.entries(manifest.schemas ?? {})) {
  const schema = contract.schemas[name];
  if (!schema) {
    fail(`schema ${name}`, "the API no longer defines it.");
    continue;
  }
  const properties = schema.properties ?? {};

  const check = (field, expectation) => {
    const property = properties[field];
    if (!property) {
      fail(
        `${name}.${field}`,
        `the API no longer has this property. It has: ${Object.keys(properties).sort().join(", ")}`,
      );
      return;
    }
    if (expectation === "required" && property.required !== true) {
      fail(
        `${name}.${field}`,
        "this client reads it as always present, but the API no longer requires it.",
      );
    }
    if (expectation === "optional" && property.required === true) {
      fail(
        `${name}.${field}`,
        "this client omits it, but the API now requires it.",
      );
    }
  };

  for (const field of declared.required ?? []) check(field, "required");
  for (const field of declared.optional ?? []) check(field, "optional");
  for (const field of declared.fields ?? []) check(field, "present");
}

// ── enum values ──────────────────────────────────────────────────────────────

for (const [path, declared] of Object.entries(manifest.enums ?? {})) {
  const [name, field] = path.split(".");
  const property = contract.schemas[name]?.properties?.[field];
  if (!property) {
    fail(`enum ${path}`, "the API no longer has this property.");
    continue;
  }
  if (!Array.isArray(property.enum)) {
    fail(`enum ${path}`, "the API no longer constrains this property to an enum.");
    continue;
  }
  const values = Array.isArray(declared) ? declared : (declared.values ?? []);
  const missing = values.filter((value) => !property.enum.includes(value));
  if (missing.length > 0) {
    fail(
      `enum ${path}`,
      `this client handles ${missing.join(", ")}, which the API no longer accepts. ` +
        `It now accepts: ${property.enum.join(", ")}`,
    );
  }
  if (!Array.isArray(declared) && declared.exhaustive === true) {
    const extra = property.enum.filter((value) => !values.includes(value));
    if (extra.length > 0) {
      fail(
        `enum ${path}`,
        `this client claims to handle every value but the API added: ${extra.join(", ")}`,
      );
    }
  }
}

// ── report ───────────────────────────────────────────────────────────────────

if (problems.length > 0) {
  console.error(`SDK contract check FAILED for ${SDK} (${problems.length} problems)\n`);
  for (const { scenario, detail } of problems) {
    console.error(`  ${scenario}`);
    console.error(`      ${detail}\n`);
  }
  console.error(
    "The server's wire contract moved. Update sdk/typescript to match, then\n" +
      "adjust sdk/contract/manifests/typescript.json. See sdk/contract/README.md.",
  );
  process.exit(1);
}

const counts =
  `${(manifest.operations ?? []).length} operations, ` +
  `${Object.keys(manifest.schemas ?? {}).length} schemas, ` +
  `${Object.keys(manifest.enums ?? {}).length} enums`;
console.log(`SDK contract check ok for ${SDK} (${counts})`);
